src/fhir/api.ts

```diff
@@ -1,31 +1,70 @@
-import {patientSchema} from "../components/reports/formDataValidation";
-import {createPatientEntry} from "./resources";
+import {Resource} from "@smile-cdr/fhirts/dist/FHIR-R4/classes/models-r4";
+import {FormValues} from "../components/reports/ReportForm";
+import {
+  furtherTestingAndId,
+  organisationAndId,
+  patientAndId, planDefinitionAndId,
+  practitionersAndIds, reportAndId, serviceRequestAndId,
+  specimenAndId,
+  variantAndId,
+} from "./resources";
 
 /**
  * Create a report bundle
- * @param patient FHIR patient entity
+ * @param form values from the front end form
  */
-export const bundleRequest = (patient: typeof patientSchema) => {
+export const bundleRequest = (form: FormValues) => {
   return {
     url: "/",
     method: "POST",
     headers: {'Content-Type': 'application/fhir+json;charset=UTF-8'},
-    body: JSON.stringify(createBundle(patient)),
+    body: JSON.stringify(createBundle(form)),
   };
 }
 
-export const createBundle = (patient: typeof patientSchema) => {
-  const mrn = patient?.identifier?.at(0)?.value;
-  const patientEntry = createPatientEntry(patient);
-
+export const createBundle = (form: FormValues) => {
+  const org = organisationAndId(form.address);
+  const patient = patientAndId(form.patient, org.id);
+  const specimen = specimenAndId(form.sample, patient.id);
+  const furtherTesting = furtherTestingAndId(form.result, patient.id);
+  const plan = planDefinitionAndId(form.sample, form.result, patient.id);
+  const {authoriser, reporter} = practitionersAndIds(form.result);
+  const variant = variantAndId(form.variant, patient.id, specimen.id, specimen.identifier, reporter.id, authoriser.id);
+  const serviceRequest = serviceRequestAndId(form.sample, patient.id, plan.id, reporter.id, specimen.id);
+  const report = reportAndId(form.result, patient.id, reporter.id, authoriser.id, org.id, specimen.id, [variant.id]);
```

Comment:
Yeah I can appreciate that - definitely if there's any further abstraction of common functions then that'd be handy to raise, but may need to reassess later. The data model currently seems to be slightly fluid so trying to pin this down so that we can be sure that we're doing this consistently